pkg/tracing/tracer.go

```diff
@@ -0,0 +1,277 @@
+// Copyright 2019 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package tracing
+
+import (
+	"sync"
+	"time"
+
+	"github.com/pingcap/errors"
+	"github.com/siddontang/go/sync2"
+	"golang.org/x/net/context"
+	"google.golang.org/grpc"
+
+	"github.com/pingcap/dm/dm/pb"
+	"github.com/pingcap/dm/pkg/log"
+)
+
+var (
+	uploadInterval = 1 * time.Minute
+)
+
+// Tracer is a tracing service provider
+type Tracer struct {
+	sync.Mutex
+	wg     sync.WaitGroup
+	closed sync2.AtomicBool
+
+	ctx    context.Context
+	cancel context.CancelFunc
+
+	cfg Config
+	cli pb.TracerClient
+
+	tso   *tsoGenerator
+	idGen *IDGenerator
+
+	jobs       map[EventType]chan *Job
+	jobsClosed sync2.AtomicBool
+	rpcWg      sync.WaitGroup
+}
+
+// NewTracer creates a new Tracer
+func NewTracer(cfg Config) *Tracer {
+	t := Tracer{
+		cfg:   cfg,
+		tso:   newTsoGenerator(),
+		idGen: NewIDGen(),
+	}
+	t.jobsClosed.Set(true)
+	t.ctx, t.cancel = context.WithCancel(context.Background())
+	return &t
+}
+
+// Enable returns whether tracing is enabled
+func (t *Tracer) Enable() bool {
+	return t.cfg.Enable
+}
+
+// Start starts tracing service
+func (t *Tracer) Start() {
+	if !t.Enable() {
+		return
+	}
+	conn, err := grpc.Dial(t.cfg.TracerAddr, grpc.WithInsecure(), grpc.WithBackoffMaxDelay(3*time.Second))
+	if err != nil {
+		log.Errorf("[tracer] grpc dial error: %s", errors.ErrorStack(err))
+		return
+	}
+	t.cli = pb.NewTracerClient(conn)
+	t.closed.Set(false)
+	t.newJobChans()
+
+	t.wg.Add(1)
+	go func() {
+		ctx2, _ := context.WithCancel(t.ctx)
+		t.tsoProcessor(ctx2)
+		// cancel()
+	}()
+
+	for _, ch := range t.jobs {
+		t.wg.Add(1)
+		go func(c <-chan *Job) {
+			ctx2, _ := context.WithCancel(t.ctx)
+			t.jobProcessor(ctx2, c)
+			// cancel()
+		}(ch)
+	}
+}
+
+// Stop stops tracer
+func (t *Tracer) Stop() {
+	if !t.Enable() {
+		return
+	}
+	t.Lock()
+	defer t.Unlock()
+	if t.closed.Get() {
+		return
+	}
+	if t.cancel != nil {
+		t.cancel()
+	}
+	t.closed.Set(true)
+	t.wg.Wait()
+}
+
+// PrepareRPC calls before each rpc request to tracing server
+func (t *Tracer) PrepareRPC() {
+	t.rpcWg.Add(1)
```

Comment:
where do we call `t.rpcWg.Wait()`?

Reply:
only used in unit test now

Reply:
PrepareRPC is removed

Reply:
will it be called later (other than in unit test)?

Reply:
if want to ensure no rpc call is pending, then it is needed